### DISCUSSION TITLE
Removing electrical delay from VNA data with .delay() - maybe bugged?

### DISCUSSION THREAD

**Discussion by ttshaw1** [scikit-rf/scikit-rf#1068] [Q&A] [answered]
I have a DUT connected to a VNA measuring S21. The VNA GUI has an electrical delay compensation that I can tweak to remove phase from the lines to the DUT, leaving only the phase response of the DUT itself. But when I save the raw data, the VNA doesn't apply the delay compensation, so I need to do it in scikit-rf.

I thought that when I create the network with the saved data, I could use Network.delay(-57.4, 'ns'). But that turns out to not remove the line phase. Oddly, I find that I can minimize the angle-frequency slope by setting delay close to 0. But the resulting angle vs. frequency plot doesn't show the features of the DUT; somehow that data's lost. It seems like instead of a delay line connected to my network, I see only the delay line.

Creating an RG-58 line with the required negative delay and cascading that with my DUT network does work as expected, though. So, am I misunderstanding what Network.delay() is supposed to do, or how to use it?

![image](https://github.com/scikit-rf/scikit-rf/assets/40922470/72759583-2874-4d39-b715-cf5c35fc78c9)

[Here's the python script I ran to generate the above plot](https://github.com/scikit-rf/scikit-rf/files/15252927/analyze_vna_data.txt) - I had to change the .py extension to .txt or Github wouldn't let me upload it. 

```python
import skrf as rf
from skrf.media import Freespace, Coaxial



data_folder = 'data'

fluxonium_0v = rf.Network(data_folder + '/dut.s2p')
# fluxonium_5v = rf.Network(data_folder + '/fluxonium_2_5v.s2p')

f_rg58 = rf.Frequency(5.5, 9.5, 100001, 'GHz')
rg58 = Coaxial(f_rg58, z0_port = 50)
delay_line = rg58.line(-57.4, unit = 'ns', name = 'RG58 Delay Line')

fluxonium_0v.plot_s_deg_unwrap(1, 0, label = 'DUT')

delayed_0v = delay_line ** fluxonium_0v
delayed_0v.plot_s_deg_unwrap(1, 0, label = '-57.4ns RG 58 Delay Cascaded')

# delayed_5v = delay_line ** fluxonium_5v
# delayed_5v.plot_s_deg(1, 0)


#for some reason this doesn't work
delayed_0v = fluxonium_0v.delay(-57.4, 'ns', media = rg58) #this comes close to a flat line. 0 is the undelayed data
delayed_0v.plot_s_deg_unwrap(1,0, label = 'DUT with -57.4ns Delay')

delayed_0v = fluxonium_0v.delay(1e-12, 'ns', media = rg58) #this comes close to a flat line. 0 is the undelayed data
delayed_0v.plot_s_deg_unwrap(1,0, label = 'DUT with 1e-12ns Delay')
```

**Comment by jhillairet** [comment]
What if you create a positive time delay line, then inversing (https://scikit-rf.readthedocs.io/en/latest/api/generated/skrf.network.Network.inv.html) it when cascading?

```python
delay_line = rg58.line(57.4, unit = 'ns', name = 'RG58 Delay Line')

fluxonium_0v.plot_s_deg_unwrap(1, 0, label = 'DUT')

delayed_0v = delay_line.inv ** fluxonium_0v
```

**Reply by ttshaw1** [reply]
That has the same effect as the negative delay line, which is expected. But I don't see that that tells me anything about why fluxonium_0v.delay() can't seem to do the same thing.

**Reply by jhillairet** [reply]
could you provide a minimal example with some test data?

**Reply by ttshaw1** [reply]
Sure. The graph is different from my screenshot above because my initial test data was a larger file than necessary. [delay_example.zip](https://github.com/scikit-rf/scikit-rf/files/15298738/delay_example.zip)

![image](https://github.com/scikit-rf/scikit-rf/assets/40922470/f0af798e-6153-4ad3-aa6b-a0a0dbd4eb43)
DUT has a phase roll we need to correct. Defining a delay line and connecting it with ** works as expected, but using dut.delay() doesn't. The 1e-12ns case is particularly interesting; instead of doing basically nothing it makes the angle basically 0 for all frequencies.

**Comment by miek** [comment]
I've put together a simple example which hopefully helps to narrow down the issue:
```python
import skrf
import matplotlib.pyplot as plt

n = 11
f = skrf.Frequency(start=1e9, stop=6e9, npoints=n, unit='Hz')
n1 = skrf.Network(frequency=f, s=[-0.9]*n, name='short')
n2 = skrf.Network(frequency=f, s=[0.9]*n, name='open')
n = skrf.two_port_reflect(n1, n2)

d1 = n.delay(30, 'ps', port=0)
d2 = n.delay(30, 'ps', port=1)

plt.subplot(121).set_title('Delay port 1')
d1.plot_s_smith(marker='.')

plt.subplot(122).set_title('Delay port 2')
d2.plot_s_smith(marker='.')

plt.show()
```

This creates a network with a short on port 1 and an open on port 2, then plots the result of applying a delay to each port:
![skrf_delay_test](https://github.com/scikit-rf/scikit-rf/assets/578095/63bfd04e-22e5-445a-811f-e3baf59b9f43)

Adding the delay to port 2 looks correct to me, but there's a problem with applying it to port 1 - it looks like the network gets flipped after applying the delay, so the (delayed) short is now on port 2. I think this explains some of the strange results seen above, since it'll end up plotting S12 which just had placeholder values in the s2p.

**Comment by mhuser** [comment, marked as answer]
Hello @ttshaw1, @jhillairet , @miek 
I looked at your examples and concluded there is a bug in `Network.delay` method.
If I do a single frequency point 2-port open and add a delay of pi rad:
```Python
import skrf as rf
import numpy as np

# n-port open
nports = 2
f = rf.Frequency(1, 1, 1, 'GHz')
s = np.identity(nports, dtype = complex)
z0 = np.array([50] * nports)
z0[0] = 25
n = rf.Network(frequency = f, s = s, z0 = z0, name = 'original')
print(n.name)
print(f"{np.angle(n.s[0])}")
print(f"z0:{n.z0[0]}")

# apply pi delay
n = n.delay(np.pi, unit = 'rad', port = 0)
n.name = 'theta = pi delay'
print(n.name)
print(f"{np.angle(n.s[0])}")
print(f"z0:{n.z0[0]}")
```
I get the following output, like @miek diagnostic:
```
original
[[0. 0.]
 [0. 0.]]
z0:[25.+0.j 50.+0.j]
theta = pi delay
[[ 0.          0.        ]
 [ 0.         -3.14159265]]
z0:[50.+0.j 25.+0.j]
```
The pi rad delay is applied but the ports are swapped (see the 25 ohm port moving from 0 to 1). This is due to `Network.connect` when cascading the delay:
https://github.com/scikit-rf/scikit-rf/blob/b67a5c2dacaf7415991c2606716ade9d5a140eab/skrf/network.py#L3426-L3427
https://github.com/scikit-rf/scikit-rf/blob/b67a5c2dacaf7415991c2606716ade9d5a140eab/skrf/network.py#L4994-L5002
Should be `if ntwkB.nports == 2 and ntwkA.nports >= 2 and num == 1:`

**Reply by mhuser** [reply]
Fixed by issue #1127 and PR #1128
